src/GafferScene/IECoreGLPreview/Renderer.cpp

```diff
@@ -525,36 +533,56 @@ class OpenGLObject : public IECoreScenePreview::Renderer::ObjectInterface
 
 				if( m_attributes->visualiserScale() > 0.0f )
 				{
-					if( haveMatchingVisualisations( Visualisation::Scale::Visualiser, categories, attrVis, m_objectVisualisations ) )
+					if( haveMatchingVisualisations( colorSpace, Visualisation::Scale::Visualiser, categories, attrVis, m_objectVisualisations ) )
 					{
 						ScopedTransform v( visualiserTransform( false ) );
-						renderMatchingVisualisations( Visualisation::Scale::Visualiser, categories, currentState, attrVis, m_objectVisualisations );
+						renderMatchingVisualisations( colorSpace, Visualisation::Scale::Visualiser, categories, currentState, attrVis, m_objectVisualisations );
 					}
 
-					if( haveMatchingVisualisations( Visualisation::Scale::LocalAndVisualiser, categories, attrVis, m_objectVisualisations ) )
+					if( haveMatchingVisualisations( colorSpace, Visualisation::Scale::LocalAndVisualiser, categories, attrVis, m_objectVisualisations ) )
 					{
 						ScopedTransform c( visualiserTransform( true ) );
-						renderMatchingVisualisations( Visualisation::Scale::LocalAndVisualiser, categories, currentState, attrVis, m_objectVisualisations );
+						renderMatchingVisualisations( colorSpace, Visualisation::Scale::LocalAndVisualiser, categories, currentState, attrVis, m_objectVisualisations );
 					}
 				}
 
-				if( haveMatchingVisualisations( Visualisation::Scale::None, categories, attrVis, m_objectVisualisations ) )
+				if( haveMatchingVisualisations( colorSpace, Visualisation::Scale::None, categories, attrVis, m_objectVisualisations ) )
 				{
 					ScopedTransform l( m_transformSansScale );
-					renderMatchingVisualisations( Visualisation::Scale::None, categories, currentState, attrVis, m_objectVisualisations );
+					renderMatchingVisualisations( colorSpace, Visualisation::Scale::None, categories, currentState, attrVis, m_objectVisualisations );
 				}
 
-				if( m_renderable || haveMatchingVisualisations( Visualisation::Scale::Local, categories, attrVis, m_objectVisualisations ) )
+				if( haveMatchingVisualisations( colorSpace, Visualisation::Scale::Local, categories, attrVis, m_objectVisualisations ) )
 				{
 					ScopedTransform l( m_transform );
-
-					renderMatchingVisualisations( Visualisation::Scale::Local, categories, currentState, attrVis, m_objectVisualisations );
-					if( m_renderable ) { m_renderable->render( currentState ); }
+					renderMatchingVisualisations( colorSpace, Visualisation::Scale::Local, categories, currentState, attrVis, m_objectVisualisations );
 				}
-
 			}
-			else if( m_renderable )
+
+			// Objects are rendered into `ColorSpace::Scene`, with the caveat that selection
+			// overlays and additional visualisations are drawn into `ColorSpace::Display`.
+
+			const IECoreGL::State *visualisationState = m_attributes->visualisationState();
+			if( m_renderable && ( colorSpace == Visualisation::ColorSpace::Scene || isSelected || visualisationState ) )
```

Comment:
In the original version, there was another path for rendering `m_renderable` inside of the `if( haveVisualisations )` branch. I believe this duplication was to avoid the construction of a second `ScopedTransform( m_transform )` in the case of having both an object _and_ a visualisation.

I removed that duplication for a few reasons :

- I concluded that `m_attributes->state()` (which comes from OpenGLAttributes nodes) shouldn't be applied to the visualisations, so needed to be bound in the `if( m_renderable )` branch only, and not upfront for everything.
- I don't think it was buying anything - I can't think of any case where we have both `m_renderable` and some visualisations.
- It was simpler.